library/rsa.c

```diff
@@ -528,16 +528,16 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
         MBEDTLS_MPI_CHK( mbedtls_mpi_gen_prime( &ctx->Q, nbits >> 1, 0,
                                                 f_rng, p_rng ) );
 
-        if( mbedtls_mpi_cmp_mpi( &ctx->P, &ctx->Q ) == 0 )
+        /* make sure the difference between p and q is not too small */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &H, &ctx->P, &ctx->Q ) );
+        if( ( nbits >= 200 ) && ( mbedtls_mpi_bitlen( &H ) < (nbits >> 1) - 100 ) )
             continue;
 
-        MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &ctx->N, &ctx->P, &ctx->Q ) );
-        if( mbedtls_mpi_bitlen( &ctx->N ) != nbits )
-            continue;
-
-        if( mbedtls_mpi_cmp_mpi( &ctx->P, &ctx->Q ) < 0 )
+        if( H.s < 0 )
```

Comment:
Where does this requirement come from?

Reply:
This just keeps the previous behavior that P > Q. If you don't think that's necessary this check can be removed.

Reply:
Oh, right - it was a pre-existing check. @mpg Do you know why we had that?

Reply:
I had no idea why, but iterated git blame led me to ef441784749d14ea9f8afad3f258489d2445c438 which in turn revealed https://github.com/ARMmbed/mbedtls/issues/558 so apparently that's why.